### DISCUSSION TITLE
Service vs EventloopService

### DISCUSSION THREAD

**Discussion by centratelemedia** [activej/activej#196] [Q&A] [answered]
Regarding Graph Service, which one should I use, service or eventloopservice, what makes them different?

**Comment by centratelemedia** [comment, marked as answer]
"/**
 * This interface is for services that can be started and then stopped
 * in the context of eventloop, so it works with {@link Promise}
 */
thank i got answer EventloopService

**Reply by eduard-vasinskyi** [reply]
Hi, @centratelemedia

In addition I’d like to clarify the difference between each service interface and when to use them:
- If your service uses [Eventloop](https://github.com/activej/activej/blob/master/core-eventloop/src/main/java/io/activej/eventloop/Eventloop.java) for asynchronous operations you can make it implement the [EventloopService](https://github.com/activej/activej/blob/master/core-promise/src/main/java/io/activej/async/service/EventloopService.java) interface. `start()`/`stop()` operations will be executed using the service's eventloop.
- If your service uses Java’s `CompletableFuture` API for asynchronous operations you can make it implement the [Service](https://github.com/activej/activej/blob/master/boot-servicegraph/src/main/java/io/activej/service/Service.java) interface. start()/stop() operations will be completed once the corresponding CompletableFutures are completed.
- If your service is synchronous you can make it implement the [BlockingService](https://github.com/activej/activej/blob/master/util-common/src/main/java/io/activej/common/service/BlockingService.java) interface. Service graph will then concurrently execute start()/stop() operations while preserving start/stop order of services.

**Reply by centratelemedia** [reply]
@eduard-vasinskyi thank for answer...
